keps/sig-cloud-provider/5237-watch-based-route-controller-reconciliation/README.md

```diff
@@ -0,0 +1,448 @@
+
+This can cause issues with limits on the number of routes in the infrastructure provider, as well as invalid routes being advertised as valid, causing possible networking reliability or confidentiality issues. s
+
+This is mitigated by:
+
+- not-filtered node event causing a full reconcile of *all* routes
+- The controller doing a periodic reconcile to clean up outdated routes, just at a lower frequency than before, even if no events came in.
+
+## Design Details
+
+We use a similar concept as already implemented in the node and service controller. Through node informers we register event handlers for the add, delete and update node events, where updates are filtered by certain criteria. To introduce the feature we establish a new feature flag called `CloudControllerManagerWatchBasedRoutesReconciliation`.
+
+#### Full reconcile
+
+The current route controller always does a “full reconcile”, looking at all routes inside the configured `ClusterCIDR` and all nodes and then taking action to create/update/delete any routes to match the expectation.
+
+This will be kept, as there is no way to know which routes to delete without this full perspective.
+
+#### Periodic reconcile
+
+As already established, the cloud-controller-manager does not utilize “Owner” references and therefore routes of already deleted nodes could remain in the infrastructure provider. To ensure a consistent state with the cluster an additional periodic reconcile should be implemented. This can be implemented via the already existing `--min-resync-period` flag, which is already used by the service and node controller.
+
+#### Workqueue singleton
+
+To ensure only a single reconciliation loop is run concurrently we use the [`k8s.io/client-go/util/workqueue`](http://k8s.io/client-go/util/workqueue) package. We only add a single key with the constant value “routes”.
+
+#### Node update filters
+
+Node updates are quite frequent. To reduce the number of requests sent to infrastructure providers, it’s important to filter these events and only trigger reconciliations when necessary.
+
+Two fields are relevant for determining whether a reconcile should occur:
+
+1. `Node.Status.Addresses` maps to the `TargetNodeAddresses` field in the `Route` struct. It determines where packets for a give IP range should be sent. Changes to this field must trigger a reconcile.
+2. `Node.Spec.PodCIDRs` contains the IP ranges assigned to the node. These CIDRs are used as the destination in the created routes.. Changes to this field must trigger a reconcile.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- Test filtering when updating a node by checking if reconcile is triggered through an item in the workqueue
+
+##### Integration tests
+
+- TBD
```

Comment:
i'm not sure if we have anywhere to do the integration testing. we might mention that integration testing needs to be undertaken by the individual cloud provider implementations.